File: src/main/java/de/flavormate/ba_entities/unit/repository/UnitConversionRepository.java
/* Licensed under AGPLv3 2024 */
package de.flavormate.ba_entities.unit.repository;

import de.flavormate.ba_entities.unit.model.UnitConversion;
import de.flavormate.ba_entities.unit.model.UnitConversionId;
import java.util.List;
import java.util.Optional;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface UnitConversionRepository extends JpaRepository<UnitConversion, UnitConversionId> {

  @Query(
      "SELECT uc.factor FROM UnitConversion uc WHERE uc.id.from.id = :from AND uc.id.to.id = :to")
  Optional<Double> findFactor(@Param("from") Long from, @Param("to") Long to);

  @Query("SELECT uc.id.to.id FROM UnitConversion uc WHERE uc.id.from.id = :from")
  List<Long> findConversions(@Param("from") Long from);
}

File: src/main/kotlin/de/flavormate/features/unit/repositories/UnitConversionRepository.kt
/* Licensed under AGPLv3 2024 - 2026 */
package de.flavormate.features.unit.repositories

import de.flavormate.features.unit.daos.models.UnitConversionEntity
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase
import jakarta.enterprise.context.ApplicationScoped

@ApplicationScoped
class UnitConversionRepository : PanacheRepositoryBase<UnitConversionEntity, String> {
  fun findFactor(source: String, target: String): Double? {
    return find("id.source.id = ?1 and id.target.id = ?2", source, target).firstResult()?.factor
  }

  fun findConversions(source: String): List<String> {
    return find(
        "select ui.id.target.id from UnitConversionEntity uc where uc.id.source.id = ?1",
        source,
      )
      .project(String::class.java)
      .list()
  }
}
